### PR TITLE
acl: RPC endpoints for JWT auth

### PR DESCRIPTION
**Pull request by pkazmierczak**
This PR introduces RPC endpoints for JWT auth.

### CODE REVIEW

File: nomad/structs/acl.go [hunk 1]
```diff
@@ -898,6 +926,90 @@ func (a *ACLAuthMethodConfig) Copy() *ACLAuthMethodConfig {
...
+		ClockSkewLeeway  any
+		*Alias
+	}{
+		Alias: (*Alias)(a),
+	}
+	if err = json.Unmarshal(data, &aux); err != nil {
+		return err
+	}
+	if aux.ExpirationLeeway != nil {
+		switch v := aux.ExpirationLeeway.(type) {
```
Comment (shoenig) [inline nomad/structs/acl.go:972]: should we have a `default` case that returns error on unexpected type?

**Review by shoenig — APPROVED** [review]
disclaimer: I'm not a JWT/auth expert

File: nomad/structs/acl.go [hunk 2]
```diff
@@ -1378,9 +1493,38 @@ func (a *ACLOIDCCompleteAuthRequest) Validate() error {
...
+
+// ACLLoginRequest is the request object to begin auth with an external bearer
+// token provider.
+type ACLLoginRequest struct {
+
+	// AuthMethodName is the name of the auth method being used to login. This
+	// is a required parameter.
+	AuthMethodName string
+
+	// LoginToken is the token used to login. This is a required parameter.
```
Comment (jrasell) [inline nomad/structs/acl.go:1511]: Would you be able to expand where this token comes from? It's not clear to me exactly where this comes from, and whether this is different from a Nomad ACL token.

File: nomad/mock/acl.go [hunk 3]
```diff
@@ -228,16 +233,17 @@ func ACLAuthMethod() *structs.ACLAuthMethod {
...
-			OIDCClientID:        "mock",
-			OIDCClientSecret:    "very secret secret",
-			OIDCScopes:          []string{"groups"},
-			BoundAudiences:      []string{"audience1", "audience2"},
-			AllowedRedirectURIs: []string{"foo", "bar"},
-			DiscoveryCaPem:      []string{"foo"},
-			SigningAlgs:         []string{"bar"},
-			ClaimMappings:       map[string]string{"foo": "bar"},
-			ListClaimMappings:   map[string]string{"foo": "bar"},
+			JWTValidationPubKeys: []string{},
```
Comment (jrasell) [inline nomad/mock/acl.go:236]: Would it be better to have separate mocks for OIDC and JWT based auth methods?

File: nomad/acl_endpoint.go [hunk 4]
```diff
@@ -2735,3 +2741,144 @@ func (a *ACL) OIDCCompleteAuth(
 	reply.ACLToken = tokenUpsertReply.Tokens[0]
 	return nil
 }
+
+// Login RPC performs non-interactive auth using a given AuthMethod. This method
+// can not be used for OIDC login flow.
+func (a *ACL) Login(args *structs.ACLLoginRequest, reply *structs.ACLCompleteAuthResponse) error {
```
Comment (jrasell) [inline nomad/acl_endpoint.go:2747]: Do we want to version gate this RPC? In a situation where one server in the region or one region in a federated cluster has not been updated to a version that includes this, I believe the call will always error.

File: lib/auth/jwt/validator.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse public key for JWT auth: %v", err)
+		}
+	}
+	return jwt.NewStaticKeySet(parsedKeys)
+}
+
+func usingJWKS(ctx context.Context, jwksurl, jwkscapem string) (jwt.KeySet, error) {
+	// Measure the JWKS endpoint performance.
+	defer metrics.MeasureSince([]string{"nomad", "acl", "jwks"}, time.Now())
```
Comment (jrasell) [inline lib/auth/jwt/validator.go:102]: What would an operator to use this to measure? I find the name clashes with RPC telemetry names and therefore initially seems an RPC measure to me.

File: lib/auth/jwt/validator.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	keySet, err := jwt.NewJSONWebKeySet(ctx, jwksurl, jwkscapem)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get validation keys from JWKS: %v", err)
+	}
+	return keySet, nil
+}
+
+func usingOIDC(ctx context.Context, oidcurl string, oidccapem []string) (jwt.KeySet, error) {
+	// Measure the OIDC endpoint performance.
+	defer metrics.MeasureSince([]string{"nomad", "acl", "oidc_jwt"}, time.Now())
```
Comment (jrasell) [inline lib/auth/jwt/validator.go:113]: What would an operator to use this to measure? I find the name clashes with RPC telemetry names and therefore initially seems an RPC measure to me.

File: lib/auth/jwt/validator.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+			)
+		}
+		var iss string
+		var ok bool
+		if iss, ok = claims["iss"].(string); !ok {
+			return nil, fmt.Errorf("unable to read iss property of provided token")
+		}
+		if !slices.Contains(methodConf.BoundIssuer, iss) {
+			return nil, fmt.Errorf("invalid JWT issuer: %v", claims["iss"])
+		}
```
Comment (jrasell) [inline lib/auth/jwt/validator.go:82]: very much a take it or leave it suggestion:
```suggestion
		if iss, ok := claims["iss"].(string); !ok {
			return nil, fmt.Errorf("unable to read iss property of provided token")
		} else if !slices.Contains(methodConf.BoundIssuer, iss) {
			return nil, fmt.Errorf("invalid JWT issuer: %v", claims["iss"])
		}
```

File: lib/auth/jwt/validator.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+		return nil, fmt.Errorf("unable to get validation keys from JWKS: %v", err)
+	}
+	return keySet, nil
+}
+
+func usingOIDC(ctx context.Context, oidcurl string, oidccapem []string) (jwt.KeySet, error) {
+	// Measure the OIDC endpoint performance.
+	defer metrics.MeasureSince([]string{"nomad", "acl", "oidc_jwt"}, time.Now())
+
+	// TODO why do we have DiscoverCaPem as an array but JWKSCaPem as a single string?
```
Comment (jrasell) [inline lib/auth/jwt/validator.go:115]: Is this something we want to figure out before merging, or is OK for a later date?

File: lib/auth/jwt/validator.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// Validate performs token signature verification and JWT header validation,
+// and returns a list of claims or an error in case any validation or signature
+// verification fails.
+func Validate(ctx context.Context, token string, methodConf *structs.ACLAuthMethodConfig) (map[string]any, error) {
+	var keySet jwt.KeySet
+	var err error
```
Comment (jrasell) [inline lib/auth/jwt/validator.go:22]: typical go style to declare multiple vars:

```suggestion
	var (
		keySet jwt.KeySet
		err    error
	)
```

**Review by jrasell — CHANGES_REQUESTED** [review]
Looking great, thanks for taking this on!  

Along with the inline comments, I feel we should expand the auth-method validate function, so that we have validation workflows for either JWT or OIDC. This will make the UX better and also removes current ambiguity in the `jwt.Validate` function whereby all three options could have been configured, but the first will only ever be used.

File: nomad/structs/acl.go [hunk 10]
```diff
@@ -1378,9 +1493,38 @@ func (a *ACLOIDCCompleteAuthRequest) Validate() error {
 	return mErr.ErrorOrNil()
 }
 
-// ACLOIDCCompleteAuthResponse is the response when the OIDC auth flow has been
+// ACLCompleteAuthResponse is the response when the auth flow has been
 // completed successfully.
-type ACLOIDCCompleteAuthResponse struct {
+type ACLCompleteAuthResponse struct {
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply nomad/structs/acl.go]: resolved in [eccc9a6](https://github.com/hashicorp/nomad/pull/15918/commits/eccc9a6941e346af13adc9303f2d2250949c0bdf)

File: nomad/mock/acl.go (hunk 3 above)
Comment (pkazmierczak) [reply nomad/mock/acl.go:265]: I think that's actually a good shout. Separated in [92ef601](https://github.com/hashicorp/nomad/pull/15918/commits/92ef60197693583d59358eae059909de0fe6c00c)

File: nomad/acl_endpoint.go (hunk 4 above)
Comment (pkazmierczak) [reply nomad/acl_endpoint.go]: Good shout, I gated to 1.5.2 with a TODO comment in the leader. See [28efef6](https://github.com/hashicorp/nomad/pull/15918/commits/28efef61a282d311e4414d37c4d387e6eacd54f8) if you're ok with the implementation.

**Review by Ayomi-gh — APPROVED** [review]
All looks good 👍

**Review by jrasell — APPROVED** [review]
LGTM; minor comment inline but that's not a blocker.

File: nomad/acl_endpoint.go [hunk 11]
```diff
@@ -2735,3 +2741,152 @@ func (a *ACL) OIDCCompleteAuth(
...
+			Region:    a.srv.Region(),
+			AuthToken: a.srv.getLeaderAcl(),
+		},
+	}
+
+	var tokenUpsertReply structs.ACLTokenUpsertResponse
+
+	if err := a.srv.RPC(structs.ACLUpsertTokensRPCMethod, &tokenUpsertRequest, &tokenUpsertReply); err != nil {
+		return err
+	}
```
Comment (tgross) [inline nomad/acl_endpoint.go:2884]: Calling this RPC will work but we'll have to go thru authentication again, check for forwarding again, write metrics as though a second RPC was called, etc. 

It'd probably be better to pull the shared bits out of the `ACLUpsertTokens` RPC into its own method and have that called from both `ACLUpsetTokens` and `ACLLogins`.

Comment (tgross) [reply nomad/acl_endpoint.go:2884]: (Fine to do in a follow-up PR)

Comment (pkazmierczak) [reply nomad/acl_endpoint.go:2884]: I'll make a PR against `main` once the feature branch is merged.

Comment (pkazmierczak) [reply nomad/acl_endpoint.go:2884]: addressed in https://github.com/hashicorp/nomad/pull/17194

**Pull request merged**